text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,629 @@
+body.
+
+### Disambiguation
+
+For the most part, paths should naturally be unique. However, there
+are some cases where a single parent may have multiple children with
+the same path. One case would be erroneous programs, where there are
+(e.g.) two structs declared with the same name in the same
+module. Another is that some items, such as impls, do not have a name,
+and hence we cannot easily distinguish them. Finally, it is possible
+to declare multiple functions with the same name within function bodies:
+
+```rust
+fn foo() {
+    {
+        fn bar() { }
+    }
+
+    {
+        fn bar() { }
+    }
+}
+```
+
+All of these cases are handled by a simple *disambiguation* mechanism.
+The idea is that we will assign a path to each item as we traverse the
+HIR. If we find that a single parent has two children with the same
+name, such as two impls, then we simply assign them unique integers in
+the order that they appear in the program text. For example, the
+following program would use the paths shown (I've elided the
+disambiguating integer except where it is relevant):
+
+```rust
+mod foo {               // Path: <root>::foo
+    pub struct Type { } // Path: <root>::foo::Type
+    impl Type {         // Path: <root>::foo::(<impl>,0)
+        fn bar() {..}   // Path: <root>::foo::(<impl>,0)::bar
+    }
+    impl Type { }       // Path: <root>::foo::(<impl>,1)
+}
+```
+
+Note that the impls were arbitrarily assigned indices based on the order
+in which they appear. This does mean that reordering impls may cause
+spurious recompilations. We can try to mitigate this somewhat by making the
+path entry for an impl include some sort of hash for its header or its contents,
+but that will be something we can add later.
+
+*Implementation note:* Refactoring DefIds in this way is a large
+task. I've made several attempts at doing it, but my latest branch
+appears to be working out (it is not yet complete). As a side benefit,
+I've uncovered a few fishy cases where we using the node id from
+external crates to index into the local crate's HIR map, which is
+certainly incorrect. --nmatsakis
+
+<a id="depgraph">
+## Identifying and tracking dependencies
+
+### Core idea: a fine-grained dependency graph
+
+Naturally any form of incremental compilation requires a detailed
+understanding of how each work item is dependent on other work items.
+This is most readily visualized as a dependency graph; the
+finer-grained the nodes and edges in this graph, the better. For example,
+consider a function `foo` that calls a function `bar`:
+
+```rust
+fn foo() {
+    ...
+    bar();
+    ...
+}
+```
+
+Now imagine that the body (but not the external signature) of `bar`
+changes. Do we need to type-check `foo` again? Of course not: `foo`
+only cares about the signature of `bar`, not its body. For the
+compiler to understand this, though, we'll need to create distinct
+graph nodes for the signature and body of each function.
```

Comment:
It doesn't really change anything here: the body has to be re-type-checked regardless, and in the case of return type inference we could simply compare the old and new inferred types. There's nothing forcing us to be conservative here.

It is true that the return type inference approach might cut down a bit on parallelism for typechecking, since it introduces sequential dependencies. It's hard to know how much, though.

In any case, FWIW I'm increasingly leaning now in the other direction, toward actual type abstraction, which would largely ameliorate these concerns.